From f02dd002cdf8d4b77b033f811639c46e39ff29b8 Mon Sep 17 00:00:00 2001
From: Jakub Filak <jfilak@redhat.com>
Date: Thu, 7 May 2015 11:07:12 +0200
Subject: [ABRT PATCH] daemon, dbus: allow only root to create CCpp, Koops,
 vmcore and xorg
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Florian Weimer <fweimer@redhat.com>:
    This prevents users from feeding things that are not actually
    coredumps and excerpts from /proc to these analyzers.

    For example, it should not be possible to trigger a rule with
    “EVENT=post-create analyzer=CCpp” using NewProblem

Related: #1212861

Signed-off-by: Jakub Filak <jfilak@redhat.com>
---
 src/daemon/abrt-server.c |  3 +++
 src/include/libabrt.h    |  2 ++
 src/lib/hooklib.c        | 24 ++++++++++++++++++++++++
 3 files changed, 29 insertions(+)

diff --git a/src/daemon/abrt-server.c b/src/daemon/abrt-server.c
index f083a1c..18f7711 100644
--- a/src/daemon/abrt-server.c
+++ b/src/daemon/abrt-server.c
@@ -443,6 +443,9 @@ static int perform_http_xact(void)
     if (repeating_crash) /* Only pretend that we saved it */
         return 0; /* ret is 0: "success" */
 
+    if (allowed_new_user_problem_entry(client_uid, FILENAME_ANALYZER, analyzer) == false)
+        return 403; /* Forbidden */
+
     return create_debug_dump();
 }
 
diff --git a/src/include/libabrt.h b/src/include/libabrt.h
index e6169b0..0af8157 100644
--- a/src/include/libabrt.h
+++ b/src/include/libabrt.h
@@ -51,6 +51,8 @@ char *run_unstrip_n(const char *dump_dir_name, unsigned timeout_sec);
 bool dir_is_in_dump_location(const char *dir_name);
 #define dir_has_correct_permissions abrt_dir_has_correct_permissions
 bool dir_has_correct_permissions(const char *dir_name);
+#define allowed_new_user_problem_entry abrt_allowed_new_user_problem_entry
+bool allowed_new_user_problem_entry(uid_t uid, const char *name, const char *value);
 
 #define g_settings_nMaxCrashReportsSize abrt_g_settings_nMaxCrashReportsSize
 extern unsigned int  g_settings_nMaxCrashReportsSize;
diff --git a/src/lib/hooklib.c b/src/lib/hooklib.c
index d8dbe20..221c2e3 100644
--- a/src/lib/hooklib.c
+++ b/src/lib/hooklib.c
@@ -287,3 +287,27 @@ bool dir_has_correct_permissions(const char *dir_name)
     }
     return true;
 }
+
+bool allowed_new_user_problem_entry(uid_t uid, const char *name, const char *value)
+{
+    /* Allow root to create everything */
+    if (uid == 0)
+        return true;
+
+    /* Permit non-root users to create everything except: analyzer and type */
+    if (strcmp(name, FILENAME_ANALYZER) != 0
+     && strcmp(name, FILENAME_TYPE) != 0
+     /* compatibility value used in abrt-server */
+     && strcmp(name, "basename") != 0)
+        return true;
+
+    /* Permit non-root users to create all types except: C/C++, Koops, vmcore and xorg */
+     if (strcmp(value, "CCpp") != 0
+      && strcmp(value, "Kerneloops") != 0
+      && strcmp(value, "vmcore") != 0
+      && strcmp(value, "xorg") != 0)
+        return true;
+
+    error_msg("Only root is permitted to create element '%s' containing '%s'", name, value);
+    return false;
+}
-- 
1.8.3.1

